### ISSUE TITLE
Rebase from upstream

### ISSUE THREAD

**Issue by pradeepto** [fabric8io/jenkins-sync-plugin#10]
Most of the implementation of downstream has done in [job-to-bc](https://github.com/fabric8io/jenkins-sync-plugin/tree/job-to-bc) branch. `job-to-bc` is a branch that is used for releasing project. 
Looks like, @rupalibehera has done a good amount of work to rebase downstream with upstream in branch [copy-job-to-bc](https://github.com/fabric8io/jenkins-sync-plugin/tree/copy-job-to-bc). 
While this rebase efforts were going on, some changes also made to `job-to-bc` simultaneously. Considering all these facts it's hard to get the last point that got synced in downstream with respect to upstream.

After having some discussion with @rupalibehera , we are thinking to rebase downstream again from upstream:master to job-to-bc branch rather than resuming the work done in `copy-job-to-bc`.

Here is the diff that need to cover https://github.com/fabric8io/jenkins-sync-plugin/compare/job-to-bc...openshift:master#diff-853317f7c9f70fec3fde84c9aeb4ec12R25 

@rupalibehera @chmouel @pradeepto WDYT?

**Comment by hrishin** [comment]
Initially, we started rebasing the code which has 54+ commits to cover. Rebasing in one go will take a lot of time and not reliable. Hence we are cherry picking one commit at a time and try to test its behavior. It looks like rebasing involves a lot of code overlapping between upstream and downstream hence it requires some methodical approach.
One thing it needs a good amount of work. Rebasing could break things, its need test suit( safety net) in place. After a call with Gabe (upstream maintainer), we realized that there is tets suit present for upstream but it's non-trivial to run [tets suit](https://github.com/openshift/jenkins-sync-plugin/tree/master/PR-Testing)  (tightly integrated into OpenShift source code).

cc: @rupalibehera @pradeepto @chmouel  @jaseemabid

**Comment by jaseemabid** [comment]
@hrishin I feel you. At this point, I'd like to see a justification for the effort you are planning to put into this. 

Do we have a 5 liner explaining what this library does for fabric8 build team as a whole? Is this really needed?

**Comment by chmouel** [comment]
it's critical for the way OpenShift and Jenkins are interacting to each others I believe, allows to have seamless integration with a Jenkins Build and an OpenShift BuildConfig, 

@hrishin what about we try in another way (if that's not how you did that before), try to use the upstream project directly and add as needed to upstream the changes needed for us, 

Is it the changes between upstream and the fork : 

https://github.com/fabric8io/jenkins-sync-plugin/compare/master...job-to-bc

The link linked earlier didn't see to have it,

**Comment by hrishin** [comment]
@jaseemabid agreed to validate the amount of work we have to do. Because this is the only bridge between Jenkins and OpenShift we are running, I'm really not sure how worth it is. 

@chmouel In my opinion current state of the downstream is good enough to work with OSIO. But tomorrow if we hit some wall then we may end up syncing the downstream with upstream again. We could try that way as well but again we have to maintain the fork. Instead of that let's bring the downstream changes into upstream and get rid of downstream maintenance once for all (For which upstream maintainer Gabe is also agreed)

WDYT?

**Comment by jaseemabid** [comment]
@chmouel has a good point here. None of us understands the changes that went to the fork very well and it's not easy doing this work on an unknown codebase. If the upstream works for our purpose, let's stick with it and throw this away. We can come back later and pick things we need. 

Classic software abandonment problems.

**Comment by chmouel** [comment]
FYI @hrishin, I am talking about the other way around, using upstream first and only cherry-pick as needed the change from downstream, so I guess we are in agreement,

I see quite a lot of changes from jstrachan in there but  I am not sure if we are those features like this :

https://github.com/fabric8io/jenkins-sync-plugin/commit/76b429fec1a506b5f5bd1f5a8b45c1e7982c28a1 <https://github.com/fabric8io/jenkins-sync-plugin/commit/76b429fec1a506b5f5bd1f5a8b45c1e7982c28a1>

using regex to match pipeline job name. I am really not sure if we are using it, but i guess that’s testable.

Do you know how can we test upstream jenkins-sync-plugin for our workflow ? It may take a while to figure out what breaking since we don’t have a comprehensive unitests

**Comment by rupalibehera** [comment]
This is critical as I see it as part of OSIO, and in the past we already faced an issue due to managing different forks. we should surely take effort to merge this one.
 I understand @hrishin your issue. I could feel during the meeting too not having much context about Jenkins workflow and trying to fix this issue can be a problem. 

I was thinking if we could squash 40+commits and make single commit of it and cherry pick that uber commit to openshift/master, I did all such random things in my last rebase too, they were around same amount of commits or more. but due that this rebase has become more complicated.

**Comment by jaseemabid** [comment]
@rupalibehera Its hard to verify such a large patch and you are throwing away history making it even harder for the next person to maintain this. We are in this mess due to lack of empathy, let's not carry it forward the same way. 

I do completely understand why you said that though. Its a hard call to make. ❤️

**Comment by hrishin** [comment]
> FYI @hrishin, I am talking about the other way around, using upstream first and only cherry-pick as needed the change from downstream, so I guess we are in agreement,

agreed @chmouel 

> Do you know how can we test upstream jenkins-sync-plugin for our workflow ? It may take a while to figure out what breaking since we don’t have a comprehensive unitests

[Tests](https://github.com/openshift/origin/blob/master/test/extended/builds/pipeline.go) has in place for upstream. But its hard, I'm trying to make it work from last two days still no success :smile:.
